ctapipe/io/zfits.py

```diff
@@ -4,109 +4,96 @@
 
 This requires the protozfitsreader python library to be installed
 """
-import logging
+from numpy import ndarray
 
-import numpy as np
-import numpy.ma as ma
-from astropy import units as u
-from astropy.time import Time
+import logging
+from ctapipe.core import Map, Field
+from ctapipe.io.containers import(
+    R0CameraContainer,
+    R0Container,
+    DataContainer
+)
+from ctapipe.io.eventsource import EventSource
 
-from ctapipe.core import Container
-from .containers import R0Container
 
 logger = logging.getLogger(__name__)
 
-try:
-    import protozfitsreader
-except ImportError as err:
-    logger.fatal(
-        "the `protozfitsreader` python module is required to access MC data: {}"
-        .format(err))
-    raise err
-
-__all__ = [
-    'zfits_event_source',
-]
-
-
-def zfits_event_source(url, max_events=None):
-    """A generator that streams data from an EventIO/HESSIO MC data file
-    (e.g. a standard CTA data file.)
-
-    Parameters
-    ----------
-    url : str
-        path to file to open
-    max_events : int, optional
-        maximum number of events to read
-
-
+__all__ = ['ZFitsEventSource']
+
+
+class ZFitsEventSource(EventSource):
+    def _generator(self):
+        from protozfitsreader import ZFile
+        for count, event in enumerate(ZFile(self.input_url)):
+            data = SST1M_DataContainer()
+            data.count = count
+            data.r0.event_id = event.event_number
+            data.r0.tels_with_data = [event.telescope_id, ]
+
+            for tel_id in data.r0.tels_with_data:
+                r0 = data.r0.tel[tel_id]
+                r0.fill_from_zfile_event(event)
+            yield data
+
+    def is_compatible(self, path):
+        from astropy.io import fits
+        try:
+            h = fits.open(path)[1].header
+        except:
+            # not even a fits file
+            return False
+
+        # is it really a proto-zfits-file?
+        return (
+            (h['XTENSION'] == 'BINTABLE') and
+            (h['EXTNAME'] == 'Events') and
+            (h['ZTABLE'] is True) and
+            (h['ORIGIN'] == 'CTA') and
+            (h['PBFHEAD'] == 'DataModel.CameraEvent')
+        )
+
+    @property
+    def is_stream(self):
```

Comment:
This EventSource reads files, no? It is not a stream. This property is used by `EventSeeker` to know if it is allowed to "seek" to an event, which is presumably allowed by this reader.

Reply:
It is a file, yes. But at the moment, the Python interface of the C++ protozfitsreader implemented [here](https://forge.in2p3.fr/projects/ctaactl/repository/show/CamerasToACTL/trunk/IO/Fits) does not expose random row access, so I assumed this means it is not compatible with the Seeker.

Reply:
Na the default of `EventSeeker` is to loop through the files to find the correct event. Only if the `EventSource` has defined a `get_event_by_index` function will the `EventSeeker` utilise a more efficient approach. So therefore `ZFitsEventSource` is compatible with `EventSeeker`, and this property should return False (or be removed, as False is the default from the parent)

Reply:
Thanks for the clarification.